src/hidden.rs

```diff
@@ -20,74 +20,253 @@
+///
+/// You can define a hidden type using any underlying type that supports `Default` and `Zeroize`.
+/// This is the case for most basic types that you probably care about.
+///
+/// You can optionally supply a label to the hidden type, which lets you define multiple hidden types that can't be
+/// interchanged. To do this, use the `hidden_label!` macro to define a label, and then use it when defining your hidden
+/// type. If you don't include a label, a default one is used for you.
+///
+/// Hidden data can't be copied, which is an intentional design decision, and you can only access it as a reference.
+/// However, it supports other functionality transparently: cloning, equality, ordering, serialization, and the like.
+/// Note an important fact about serialization: it's transparent, so the type label isn't serialized.
+/// This means you should be sure to know the label you intend during deserialization!
+///
+/// ```edition2018
+/// # use rand::rngs::OsRng;
+/// # use rand::RngCore;
+/// # use tari_utilities::hidden::{Hidden, HiddenLabel};
+/// # use tari_utilities::hidden_label;
+///
+/// // In this example, we need to handle keys for a cipher that are `[u8; 32]` byte arrays
+///
+/// // Define a label for the new hidden type
+/// // This ensures that any other `[u8; 32]` hidden types can't get confused with this one
+/// hidden_label!(CipherKeyLabel);
+///
+/// // Alias a new hidden type using the data type and label
+/// type CipherKey = Hidden<[u8; 32], CipherKeyLabel>;
+///
+/// // We can create hidden data from existing data; in this case, it's the caller's responsibility to make sure the existing data is handled securely
+/// let key_from_data = CipherKey::hide([1u8; 32]);
+///
+/// // We can access the hidden data as a reference, but not take ownership of it
+/// assert_eq!(key_from_data.reveal(), &[1u8; 32]);
+///
+/// // We can create default hidden data and then modify it as a mutable reference; this is common for functions that act on data in place
+/// let mut rng = OsRng;
+/// let mut key_in_place = CipherKey::default();
+/// rng.fill_bytes(key_in_place.reveal_mut());
+///
+/// // Cloning is safe to do
+/// let clone = key_in_place.clone();
+/// assert_eq!(key_in_place.reveal(), clone.reveal());
+/// ```
+///
+/// Type enforcement means the compiler won't let you mix and match hidden types with different labels.
+/// This can be important when dealing with different types of key material, for example.
+///
+/// ```compile_fail
+/// # use tari_utilities::hidden::{Hidden, HiddenLabel};
+/// # use tari_utilities::hidden_label;
+///
+/// // These are labels for different hidden types we want to create; in this case, keys for two different ciphers
+/// hidden_label!(KeyForCipherALabel);
+/// hidden_label!(KeyForCipherBLabel);
+///
+/// // Both ciphers have keys that are `[u8; 32]` under the hood
+/// type KeyForCipherA = Hidden<[u8; 32], KeyForCipherALabel);
+/// type KeyForCipherB = Hidden<[u8; 32], KeyForCipherBLabel);
```

Comment:
You're right, and this is a bad test. The type differentiation really matters for things like function definitions, where it shouldn't be possible for a developer to build functions operating on the wrong types of keys. Fixed in a new commit.

Having reference access to the underlying type directly (in this case, byte arrays) is still important. A major use case for mutable references is allowing KDFs and random number generators to build their output directly in the wrapper, rather than make copies that the caller needs to safely handle itself. If all the underlying types are custom (like `pub struct CipherA([u8; 32])`), then you need to write custom `reveal` and `reveal_mut` that end up doing the same thing: providing access to the byte array. Providing reference access only to the struct stops KDF or random derivation from working the way we'd want.

Reply:
@sdbondi  brings a good point to the table. Not adding the 'separation label' simplifies the use of `Hidden` (which, for the most part could be used in the future mainly with default label). Moreover, the crucial thing is not to reuse a key twice, in different contexts, which is related to the underlying data content and not so much about the type in itself. I really liked the way to add type differentiation, from a formal viewpoint, but honestly, given the complexity of the overall codebase and all the maintaining requisites, I think @sdbondi  suggestion is a very sensible one. I am happy with leaving the type separation behind, as well.